### PR TITLE
11987: Prevent Reloading the whole table upon Product Clone

### LINKED ISSUE #11987: [BUU] Don't reload whole table when cloning products
When performing a clone from the actions menu, the whole table is reloaded. This can take some time, and is jarring because you can lose your position in a long list of products. 
The old products screen does this so we need to support it here too.

Screenshot 2023-12-28 at 14 08 16

1. After confirming, attempt to clone the product using existing behaviour (the name has "COPY" prepended to it).
2. If clone succeeds, insert the row(s) directly below the cloned product.
   * Try using a "slide in" animation,  to avoid the jump which would cause the user to lose their place.
3. If the action fails, show a flash message explaining why, without changing any other part of the page.

Other notes:
* Should the pagination info be updated ("showing x of x")? 
    For the first iteration, make no changes, this will be re-evaluated later
* Any unsaved edits will remain on the page and not affect progress. ([fixes issues when other rows have been edited](https://github.com/openfoodfoundation/openfoodnetwork/pull/11846#discussion_r1433282835) as [pictured](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/c27a187e-f86d-47a8-b231-8b493b4a1c41))
   * If the product to be cloned has unsaved changes, should they be saved and included in the clone?
       We'll keep it simple and won't attempt that. As usual, unsaved changes (yellow border) are unsaved until the save button is pressed.

- Depends on #12361 
- Very similar to #12398

### PR DESCRIPTION
**Pull request by chahmedejaz**
#### What? Why?

- Closes #11987

- This PR prevents the whole table from reloading when the product is cloned.
- This means other product changes won't lost if the product is cloned.
-  i.e. If you modify Product-A, and without saving it, you cloned Product-B, Product-A changes should remain intact.
- Along with this, after cloning, the cloned product will now be sliding into the table from the right.

#### What should we test?
- The Product clone functionality
- The product modification functionality as explained above
- Try to modify some fields for the cloned product and save it

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

### CODE REVIEW

**Comment by dacook** [comment]
Great progress. I noticed something, which you might already be aware of: the CSS animation causes a problem with the dropdowns:
Screenshot 2024-06-18 at 10 14 09 am

It appears to be related to the `translateX`, but I don't know how to fix it. Maybe the final keyframe styles could be adjusted for that, or maybe we need JS to remove the class after the animation is complete..

**Comment by chahmedejaz** [comment]
> Great progress. I noticed something, which you might already be aware of: the CSS animation causes a problem with the dropdowns: Screenshot 2024-06-18 at 10 14 09 am
> 
> It appears to be related to the `translateX`, but I don't know how to fix it. Maybe the final keyframe styles could be adjusted for that, or maybe we need JS to remove the class after the animation is complete..

Yes David, this was an issue. It's being fixed now. Thanks :)

File: app/views/admin/products_v3/clone.turbo_stream.haml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+- form_with url: bulk_update_admin_products_path, method: :post, id: "products-form",
+            builder: BulkFormBuilder,
+            html: { data: { 'turbo-frame': "_self",
+              controller: "bulk-form",
+              'bulk-form-disable-selector-value': "#sort,#filters",
+              'bulk-form-error-value': defined?(@error_counts),
+              } } do |form|
+    - product_body = render partial: 'product_variant_row', locals: { form:, product: @cloned_product, product_index: @product_index, producer_options: @producer_options, category_options: @category_options, tax_category_options: @tax_category_options }, formats: :html
+= turbo_stream.before dom_id(@product) do
+    = product_body
```
Comment (mkllnk) [inline app/views/admin/products_v3/clone.turbo_stream.haml]: That's clever to save the rendered partial without the form around it.

File: app/views/admin/products_v3/_product_variant_row.html.haml [hunk 2]
```diff
@@ -1,7 +1,8 @@
 = form.fields_for("products", product, index: product_index) do |product_form|
   %tbody.relaxed{ id: dom_id(product), data: { 'record-id': product_form.object.id,
                           controller: "nested-form product",
-                          action: 'rails-nested-form:add->bulk-form#registerElements rails-nested-form:remove->bulk-form#toggleFormChanged' } }
+                          action: 'rails-nested-form:add->bulk-form#registerElements rails-nested-form:remove->bulk-form#toggleFormChanged' },
+                  class: (defined?(should_slide_in) && should_slide_in) ? 'slide-in' : '' }
```
Comment (mkllnk) [inline app/views/admin/products_v3/_product_variant_row.html.haml:5]: If you rely on an optional variable then it's good to declare that at the top of the partial:
```haml
- should_slide_in ||= false
```

That makes the interface clearer and the code easier to read.

**Review by mkllnk — APPROVED** [review]
Looking great!

I'm missing some knowledge around the BUU code to comment on some details but David will look at this on Monday as well.

File: app/views/admin/products_v3/_product_variant_row.html.haml (hunk 2 above)
Comment (chahmedejaz) [reply app/views/admin/products_v3/_product_variant_row.html.haml:5]: That's nice. Sure, I'll incorporate it. Thanks 🙂

File: app/webpacker/controllers/bulk_form_controller.js [hunk 3]
```diff
@@ -161,4 +163,34 @@ export default class BulkFormController extends Controller {
...
+        }
+      }
+    });
+
+    const productsTable = document.querySelector('.products');
+    // Above mutation function will trigger,
+    // whenever +products+ table rows (first level children) are mutated i.e. added or removed
+    // right now we are using this for product clone 
+    this.productsTableObserver.observe(productsTable, { childList: true });
+  }
```
Comment (dacook) [inline app/webpacker/controllers/bulk_form_controller.js:195]: Nice work, I like that the code is building upon basic knowledge of the DOM (rows can be added/removed, and we need to handle them somehow. The `.products` selector could probably be simply `table`, which would make it not product-specific.

However the BulkFormController, is getting quite large, and the new feature doesn't seem core to the goal of:
```
// Manage "changed" state for a form with multiple records
```

So was thinking it would be great to split it into a separate controller for handling this. We could use  a Stimulus Outlet to trigger the registerElement action (or there might be a better way). I think it might almost be as simple as:

clone.turbo_stream:
```html
%tbody{ data-controller="cloned-resource" data-cloned-resource-outlet="products-form" }
```

```javascript
class ClonedResource extends Controller {
connect() {
   bulkFormOutlet.registerElements();

  this.addEventListener('animationend', this.#removeAnimationClasses.bind(this)); // assuming controller is initialised before animation finishes. But that's probably a safe assumption
}
...
```

Smaller modules will probably help with maintenance and re-use in the future, but maybe it risks creating a hug sea of controllers that is hard to navigate.. 

Would you be willing to give it a try?

File: app/controllers/admin/products_v3_controller.rb [hunk 4]
```diff
@@ -68,6 +68,28 @@ def destroy_variant
       end
     end
 
+    def clone
+      @product = Spree::Product.find(params[:id])
```
Comment (dacook) [inline app/controllers/admin/products_v3_controller.rb:72]: I think we could be using Admin::ResourceController to automatically `load_resource` for us. But that might bring new challenges... let's not bother right now.

File: app/views/admin/products_v3/_product_row.html.haml [hunk 5]
```diff
@@ -40,7 +40,7 @@
 %td.align-right
   = render(VerticalEllipsisMenu::Component.new) do
     = link_to t('admin.products_page.actions.edit'), edit_admin_product_path(product), 'data-turbo': false
-    = link_to t('admin.products_page.actions.clone'), clone_admin_product_path(product), 'data-turbo': false
+    = link_to t('admin.products_page.actions.clone'), admin_clone_product_path(product), 'data-turbo-method': :post
```
Comment (dacook) [inline app/views/admin/products_v3/_product_row.html.haml:43]: This _looks_ simple, but I'm wondering how hard it was to find the exact right name for `'data-turbo-method': :post`? Maybe we need to document this somewhere 🤔

File: app/webpacker/controllers/bulk_form_controller.js (hunk 3 above)
Comment (dacook) [reply app/webpacker/controllers/bulk_form_controller.js:195]: Sigh, putting on a project manager hat 🤠 , I realise that we need more focus on just getting this released. I think the current solution is fine, so I would suggest we can merge this. 
And if you would like to, we can deal with this as a tech debt issue later.

**Review by dacook — APPROVED** [review]
Great work, thanks for working through the challenges, you solved it!! 🎉 

It caused me to think of more ideas for refactoring, but I think we should proceed with it it like this. I see that you were going to add a small update as suggested by Maikel: would you be able to add that soon? If not, just comment that this is ready for testing.

**Comment by chahmedejaz** [comment]
> It caused me to think of more ideas for refactoring, but I think we should proceed with it it like this.

Agreed. Yes, we can incorporate these ideas in a separate tech debt :)

>I see that you were going to add a small update as suggested by Maikel: would you be able to add that soon? If not, just comment that this is ready for testing.

Ah, really sorry, that was a miss on my part 😅. If I make any changes now, then we would again need 2 approvals 😅. I think this should be tested as soon as possible so that we can know if there are any concerning issues in it. If we find anything in testing, then we can incorporate the suggestions as well, no worries. 
Otherwise, there are tech debt opportunities as well, we can note these for them as well. Thanks

This is ready for testing. Thanks

File: app/views/admin/products_v3/clone.turbo_stream.haml (hunk 1 above)
Comment (chahmedejaz) [reply app/views/admin/products_v3/clone.turbo_stream.haml]: > If I remember rightly, this was because of the weird repeating issue. It would be good to have a code comment here to explain why we do it this way.

Yes, @dacook. However, this was the issue when we were passing the controller's `view_context` to create the form and then use/pass it in the partials. That was also kind of violating the view controller separations of concerns.
Now we are creating the form in the view itself, hence that repeating issue was fixed. 

**Just some more context here:**

We are using `-` for the form because we don't want the form element here. We are using the view's helper, creating a form object, and passing it to the partial in the view.
This was the major challenge/takeaway from this: avoid using the controller's `view_context` to create the view's elements and pass them in views to use. Just create a wrapper view, create that view thing you wanted from the controller, and that should be working as expected as we are just using view for view's thing. 
The root cause might be this: 
https://www.loom.com/share/271cba54a7704572ac283535184ef78b

Please let me know if you have any questions. Thanks 🙂

File: app/webpacker/controllers/bulk_form_controller.js (hunk 3 above)
Comment (chahmedejaz) [reply app/webpacker/controllers/bulk_form_controller.js:195]: That's a great idea. Sure, I'd love to work on this in a tech debt. It would be fun 😀

Comment (chahmedejaz) [reply app/webpacker/controllers/bulk_form_controller.js:195]: Your suggestion is way better than me using a mutation observer😂. I just realized.

File: app/views/admin/products_v3/_product_row.html.haml (hunk 5 above)
Comment (chahmedejaz) [reply app/views/admin/products_v3/_product_row.html.haml:43]: Yes, this was the only stimulus-related takeaway from this. Sure, will document this one 👍🏻

**Comment by drummer83** [comment]
Hi @chahmedejaz and BUU-team (@RachL, @dacook),

First, thanks for the hard work Ahmed! It's very much appreciated! :pray: 

I have tested this PR and the main functionality is working really well. However, I found quite a few discrepancies between specification and implementation. Also I could trigger an error 500. Please have a look below and comment on how to proceed.

Thanks!

## Before
- Cloned product appears in the list in alphabetical order (COPY OF...).
- The whole page is being reloaded.

## After
### Successes
- Products can be cloned. :heavy_check_mark: 
- No reload of whole page. :heavy_check_mark: 
- Keep unsaved changes of other products. :heavy_check_mark: 
- Unsaved changes of the cloned product are not cloned but kept. :heavy_check_mark: 
- Dropdowns are shown properly. :heavy_check_mark: 

### Deviations
- Cloned product appears below the cloned product (COPY OF...). :x: It's appearing above.
- Display with slide in transition. :x: Yes, but a sliding from above would be more intuitive. (see [here](https://openfoodnetwork.slack.com/archives/C01CXQNJ1J6/p1719208519616359?thread_ts=1718874365.073349&cid=C01CXQNJ1J6)).
- Grey out the pagination info after cloning because it's outdated. :x:

### Unclear
- On failing to clone: show a flash message explaining why, without changing any other part of the page. :question: How can I make cloning fail?

 ### Fails
- Enable saving of unsaved changes after cloning. :x: Error 500 (see below).

## Issues
### Unsaved changes can't be saved after cloning
- Make changes to any product or variant without saving, then clone a product, then try to save the changes --> error 500
image
```
XHRPOST
https://staging.openfoodnetwork.org.au/admin/products/bulk_update
[HTTP/2 500  1329ms]
```

### Out of scope: Clicking on last page after deleting products shows 'no products'
- Delete products until the number of pages required for display is reduced, e.g. with '15 per page' make sure you have 16 products (two pages), then delete one product, then click on page '2' --> 'no products'
It may seem like ALL products have been deleted (scary) and only a page reload brings them back.
image

**Comment by RachL** [comment]
Thanks @drummer83 🙏 Let's fix the error 500 as part of this PR and open a new issues to fix the deviation section and the issue when deleting the last product on the page (I wonder how the legacy page deals with it!). this PR is quite big already :) 

Regarding how to make the cloning process fail - I must admit I have no idea :/

**Comment by drummer83** [comment]
Back to 'In Progress' then. Thanks!

**Comment by chahmedejaz** [comment]
Thank you @drummer83  for triggering the 500. Let me fix it and push the fix. :)

**Comment by chahmedejaz** [comment]
@drummer83  -  I tried to produce the error, but it seems to be working fine. 
https://www.loom.com/share/2d955a8d713b4badade9f170cf68b6cc

Please let me know if I'm missing something here. Thanks.

**Comment by dacook** [comment]
I think the problem was always with trying to update the second item in the list, after cloning. This is because it has an index of  "1" in the background, and the newly cloned product also had that index. 

@chahmedejaz I've pushed up a new commit which I think should solve it.

File: app/controllers/admin/products_v3_controller.rb [hunk 6]
```diff
@@ -68,6 +68,29 @@ def destroy_variant
...
+
+      begin
+        @cloned_product = @product.duplicate
+        flash.now[:success] = t('.success')
+
+        @product_index = "-#{@cloned_product.id}"
+        @producer_options = producers
+        @category_options = categories
+        @tax_category_options = tax_category_options
+      rescue StandardError => _e
```
Comment (dacook) [inline app/controllers/admin/products_v3_controller.rb:83]: I somehow missed this in my last review. We need to avoid rescuing `StandardError` because it could be hiding other important errors. Instead we can rescue a specific error.

And we should have a test to cover the error case too. Would you be able to add one?

If the `.duplicate` method returns a false on error, perhaps we should branch off that (that's what we normally do when creating/updating in controllers). Alternatively, you could change it to rescue the specific error.

**Review by dacook — CHANGES_REQUESTED** [review]
Hi Ahmed, I noticed something else that I think we should change too

File: app/controllers/admin/products_v3_controller.rb (hunk 6 above)
Comment (chahmedejaz) [reply app/controllers/admin/products_v3_controller.rb:83]: @dacook - The duplicate method is saving the cloned product with `.save!`. I used the `StandardError`, to rescue for that the failure of `.save!` 😅. After saving, it returns the product:

https://github.com/chahmedejaz/openfoodnetwork/blob/3afb3f5ba05b7565dd32c68b638d152ee2690ec4/lib/spree/core/product_duplicator.rb#L12-L16

I think branching off would be better to handle this case, but that would cause changes in multiple places. Is that okay?

> And we should have a test to cover the error case too. Would you be able to add one?

Sure, I'll add that no worries.

**Comment by chahmedejaz** [comment]
> I think the problem was always with trying to update the second item in the list, after cloning. This is because it has an index of "1" in the background, and the newly cloned product also had that index.
> 
>  I've pushed up a new commit which I think should solve it.

Yes, @dacook - I also doubted the index, and wasn't sure if we should change the bulk update part in this one. Moreover, it seemed to be working fine, so I thought if it caused an issue we might be able to catch it in the testing. I think we did 😃
Thanks for the help here 👍🏻

File: spec/models/spree/product_spec.rb [hunk 7]
```diff
@@ -11,10 +11,18 @@ module Spree
...
           clone = product.duplicate
+
+          expect(clone).to be_persisted
           expect(clone.name).to eq 'COPY OF ' + product.name
           expect(clone.sku).to eq ""
           expect(clone.image).to eq product.image
         end
+
+        it 'fails to duplicate the product' do
+          pending 'test for all the use cases by which the +product.save+ would return false.'
```
Comment (chahmedejaz) [inline spec/models/spree/product_spec.rb:22]: It's a very rare edge case because:
a) The product we will be cloning must already have been passed through the saving process, the `new_product.save` most probably will also pass as well.
b) The case in which it could fail I think would be for very old products i.e. Now in the present new validations might have been added in the model, and cloning those products might result in this case due to failing validations. So, in that case `.save` would return false. And that's what I'm assuming is the failure system spec for the product clone.

Maybe I'm missing something here, so in that case, I've created a pending spec, on which we can do futhur spike in a separate issue.
Please let me know what you think about it. Thanks.

File: spec/system/admin/products_v3/products_spec.rb [hunk 8]
```diff
@@ -1213,6 +1209,21 @@ def expect_other_columns_visible
             expect(input_content).to match /COPY OF Apples/
           end
         end
+
+        it "fails to clone the product on page when clicked on the cloned option" do
+          # Mock the +save+ method to return fail. That's the only expected fail case
+          allow_any_instance_of(Spree::Product).to receive(:save).and_return(false)
```
Comment (chahmedejaz) [inline spec/system/admin/products_v3/products_spec.rb:1215]: Context for this here:
https://github.com/openfoodfoundation/openfoodnetwork/pull/12565#discussion_r1654688954

File: app/controllers/admin/products_v3_controller.rb (hunk 6 above)
Comment (dacook) [reply app/controllers/admin/products_v3_controller.rb]: Hi @chahmedejaz , thanks for the quick turnaround. I discussed with Maikel and we agreed it was heading in a good direction, but decided to first update specs to understand the error case better.
After doing that, it was clearer and I found that an exception could still occur due to invalid data on the variants. 

It looked difficult to handle that, so I decided it's simpler to keep the exception handling as it is, but with more specific exceptions. Does that look ok to you?

**Comment by dacook** [comment]
@mkllnk as discussed would you mind reviewing the latest change, which is summarised in: https://github.com/openfoodfoundation/openfoodnetwork/pull/12565/files/0001ffa9702d60e117e24304e4c34de39dde9cec..d7992304407306b80c5b415087e963889b02d551?w=1
If you're happy, we can add back to the testing queue.

**Review by mkllnk — APPROVED** [review]
That's good, too.

File: app/controllers/admin/products_v3_controller.rb (hunk 6 above)
Comment (chahmedejaz) [reply app/controllers/admin/products_v3_controller.rb]: Looks good @dacook - Thank you so much for the help here ⭐🙂

**Comment by chahmedejaz** [comment]
@drummer83 - It's ready for testing. Please validate if the 500 error gets fixed, along with all the previously passing scenarios. Thanks.

**Comment by drummer83** [comment]
Hi @chahmedejaz,

Thanks, here we go again!

## Before
- Cloned product appears in the list in alphabetical order (COPY OF...).
- The whole page is being reloaded.

## After
### Successes
- Products can be cloned. :heavy_check_mark: 
- No reload of whole page. :heavy_check_mark: 
- Keep unsaved changes of other products. :heavy_check_mark: 
- Unsaved changes of the cloned product are not cloned but kept. :heavy_check_mark: 
- Dropdowns are shown properly. :heavy_check_mark: 
- NEW: Enable saving of unsaved changes after cloning. :heavy_check_mark: 

### Deviations
- Cloned product appears below the cloned product (COPY OF...). :x: It's appearing above.
- Display with slide in transition. :x: Yes, but a sliding from above would be more intuitive. (see [here](https://openfoodnetwork.slack.com/archives/C01CXQNJ1J6/p1719208519616359?thread_ts=1718874365.073349&cid=C01CXQNJ1J6)).
- Grey out the pagination info after cloning because it's outdated. :x:
- On failing to clone: show a flash message explaining why, without changing any other part of the page. :x: It's not explained WHY.
image

## Out of scope
### Clicking on last page after deleting products shows 'no products'
- Delete products until the number of pages required for display is reduced, e.g. with '15 per page' make sure you have 16 products (two pages), then delete one product, then click on page '2' --> 'no products'
It may seem like ALL products have been deleted (scary) and only a page reload brings them back.
image

### Page buttons not working properly
- Viewing page 3 of 3 in the product list, making an update without saving, then clicking on "2" to view page 2 of 3, takes me to page "1".

[Kazam_screencast_00000.webm](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/49e4b1a5-1c11-4c8d-a738-11af50b937fa)

### Scroll to top of the list after switching pages
- Switching to another page of the list should scroll to the top of the list.

# Conclusion
The error 500 is gone now. Deviations and issues out of scope can be handled in separate issues and PRs. 
This on is ready to go! :muscle: 
Thanks again, Ahmed! :pray:

**Comment by drummer83** [comment]
@chahmedejaz 
I am creating the issues just now! :+1:
